### PR TITLE
Internal Improvement: Port truffle-artifactor to TS

### PR DESCRIPTION
**Pull request by CruzMolina**
This PR ports the `truffle-artifactor` library (excluding tests) to TypeScript.

Adds preliminary typings to `truffle-contract` & `truffle-contract-schema`.

### CODE REVIEW

File: packages/truffle-contract-schema/typings/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+declare module "truffle-contract-schema" {
+  import Contract from "truffle-contract";
+  namespace Schema {
+    export function normalize(dirtyObj: object, opts?: object): Contract.TruffleContract;
```
Comment (gnidan) [inline packages/truffle-contract-schema/typings/index.d.ts]: This returns the ContractObject, not a truffle-contract Contract...

i.e. this is more accurate:
```typescript
import { ContractObject } from "truffle-contract-schema/spec";
```

File: packages/truffle-artifactor/.gitignore [hunk 2]
```diff
@@ -4,3 +4,4 @@ config/development
 config/test
 yarn.lock
 package-lock.json
+dist
```
Comment (gnidan) [inline packages/truffle-artifactor/.gitignore]: I don't think these nested .gitignores do anything.

File: packages/truffle-artifactor/README.md [hunk 3]
```diff
@@ -3,7 +3,7 @@
 This package saves contract artifacts into JSON files
 
 ```javascript
-const Artifactor = require("truffle-artifactor");
+const { Artifactor } = require("truffle-artifactor");
```
Comment (gnidan) [inline packages/truffle-artifactor/README.md]: Breaking change?

File: packages/truffle-artifactor/artifactor.ts [hunk 4]
```diff
@@ -1,15 +1,16 @@
...
-class Artifactor {
-  constructor(destination) {
+export class Artifactor {
+  destination: string;
+  constructor(destination: string) {
     this.destination = destination;
   }
 
-  async save(artifactObject) {
+  async save(artifactObject: object) {
```
Comment (gnidan) [inline packages/truffle-artifactor/artifactor.ts]: We can't type this any more specifically than `object`? It should be a truffle-contract Contract

File: packages/truffle-artifactor/utils.ts [hunk 5]
```diff
@@ -1,19 +1,23 @@
-const fse = require("fs-extra");
-const _ = require("lodash");
+import fse from "fs-extra";
+import _ from "lodash";
+import Contract from "truffle-contract";
 
-const writeArtifact = (completeArtifact, outputPath) => {
+export function writeArtifact(
+  completeArtifact: Contract.TruffleContract,
```
Comment (gnidan) [inline packages/truffle-artifactor/utils.ts]: This should be a ContractObject from `truffle-contract-schema/spec`

File: packages/truffle-artifactor/utils.ts [hunk 6]
```diff
@@ -1,19 +1,23 @@
...
-};
+}
 
-const finalizeArtifact = (
-  normalizedExistingArtifact,
-  normalizedNewArtifact
-) => {
+export function finalizeArtifact(
+  normalizedExistingArtifact: Contract.TruffleContract,
+  normalizedNewArtifact: Contract.TruffleContract
```
Comment (gnidan) [inline packages/truffle-artifactor/utils.ts]: Same with these

File: packages/truffle-contract-schema/.gitignore [hunk 7]
```diff
@@ -1,4 +1,4 @@
 node_modules
 yarn.lock
 package-lock.json
-*.d.ts
\ No newline at end of file
+spec/index.d.ts
```
Comment (gnidan) [inline packages/truffle-contract-schema/.gitignore]: (Same inquiry about nested .gitignores)

File: packages/truffle-contract-schema/package.json [hunk 8]
```diff
@@ -3,6 +3,7 @@
   "version": "3.0.11",
   "description": "JSON schema for contract artifacts",
   "main": "index.js",
+  "typings": "./typings/index.d.ts",
```
Comment (gnidan) [inline packages/truffle-contract-schema/package.json]: Does this conflict with the types exposed by `truffle-contract-schema/spec`?

File: packages/truffle-contract/typings/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+declare module "truffle-contract" {
+  namespace TruffleContract {
+    export interface TruffleContract {
+      contractName: string;
+      contract_name: string;
```
Comment (gnidan) [inline packages/truffle-contract/typings/index.d.ts]: blech, both!?

File: packages/truffle-core/test/config.js [hunk 10]
```diff
@@ -6,7 +6,7 @@ var Contracts = require("truffle-workflow-compile");
 var Ganache = require("ganache-core");
 var provision = require("truffle-provisioner");
 var Resolver = require("truffle-resolver");
-var Artifactor = require("truffle-artifactor");
+const { Artifactor } = require("truffle-artifactor");
```
Comment (gnidan) [inline packages/truffle-core/test/config.js]: The more I see this, the more I think we should just keep our `index.js` to maintain the default export, where all that file would do is import from `./dist`

Comment (CruzMolina) [reply packages/truffle-core/test/config.js]: Is there a way to `export default` w/o breaking things? Importing destructured like this seemed prettier than having to do `const Artifactor = require("truffle-artifactor").default` everywhere.

File: packages/truffle-artifactor/.gitignore (hunk 2 above)
Comment (CruzMolina) [reply packages/truffle-artifactor/.gitignore]: I can remove them, but they do seem to make git ignore them when working locally.

File: packages/truffle-contract-schema/.gitignore (hunk 7 above)
Comment (CruzMolina) [reply packages/truffle-contract-schema/.gitignore]: Same here. This was making git ignore the added typings, so I set it to only ignore the output spec typings. (Looks like the pkg was setup to ignore the spec typings on purpose).

File: packages/truffle-artifactor/artifactor.ts (hunk 4 above)
Comment (CruzMolina) [reply packages/truffle-artifactor/artifactor.ts]: I think `artifactObject` is technically not a `ContractObject` until we normalize it. I believe we allow passing any object in there. Maybe that doesn't matter and I should just type it as a `ContractObject`?

File: packages/truffle-contract/typings/index.d.ts (hunk 9 above)
Comment (CruzMolina) [reply packages/truffle-contract/typings/index.d.ts]: yuppp, I believe we currently support both.

File: packages/truffle-contract-schema/package.json (hunk 8 above)
Comment (CruzMolina) [reply packages/truffle-contract-schema/package.json]: For whatever reason, we don't currently expose those typings. Should I do so within `typings/index.d.ts`?

File: packages/truffle-artifactor/.gitignore (hunk 2 above)
Comment (gnidan) [reply packages/truffle-artifactor/.gitignore]: Oh, well, empirical observation is better than my postulatin'!

File: packages/truffle-artifactor/artifactor.ts (hunk 4 above)
Comment (gnidan) [reply packages/truffle-artifactor/artifactor.ts]: Nope, you're probably right! Leave it!

File: packages/truffle-contract-schema/package.json (hunk 8 above)
Comment (gnidan) [reply packages/truffle-contract-schema/package.json]: Yep, I think so!

File: packages/truffle-contract/typings/index.d.ts (hunk 9 above)
Comment (gnidan) [reply packages/truffle-contract/typings/index.d.ts]: _sigh_ ok

File: packages/truffle-core/test/config.js (hunk 10 above)
Comment (gnidan) [reply packages/truffle-core/test/config.js]: You have to do it in JS I believe

Comment (gnidan) [reply packages/truffle-core/test/config.js]: Or maybe you can finagle the tsconfig to make it work the way you want.

File: packages/truffle-contract-schema/spec/contract-object.spec.json [hunk 11]
```diff
@@ -11,6 +11,11 @@
       "description": "Name used to identify the contract",
       "default": "Contract"
     },
+    "contract_name": {
+      "allOf": [{ "$ref": "#/definitions/ContractName" }],
+      "description": "Name used to identify the contract",
+      "default": "Contract"
+    },
```
Comment (gnidan) [inline packages/truffle-contract-schema/spec/contract-object.spec.json]: Blech, can we not do this? (Is it possible?)

Comment (CruzMolina) [reply packages/truffle-contract-schema/spec/contract-object.spec.json]: i believe changing this [line](https://github.com/trufflesuite/truffle/pull/2220/files#diff-1e819f795cdbdcfa8915a5e598b511b2R49) to use `contractName` would be a breaking change 😞 .

Comment (gnidan) [reply packages/truffle-contract-schema/spec/contract-object.spec.json]: Shouldn't be breaking. That's what `normalize` is for... `contractName` is the schema-correct version, and `contract_name` should get converted via normalize.

**Pull request merged**